### PR TITLE
Check for llama_get_logits_ith() errors

### PR DESCRIPTION
**Pull request by jart**
Embeddings models like BERT don't have logits. This caused the llamafile software to crash for users who tried to inference mxbai-embed-large-v1. This change potentially helps prevent the server from crashing. Since it is possible for this function to fail having callers check the result is a good idea from a defensive coding standpoint. The older exception code has also been refactored, since it's no longer needed.

### CODE REVIEW

File: examples/perplexity/perplexity.cpp [hunk 1]
```diff
@@ -638,6 +638,9 @@ static results_perplexity perplexity(llama_context * ctx, const gpt_params & par
 
         for (int seq = 0; seq < n_seq_batch; seq++) {
             const float * all_logits = num_batches > 1 ? logits.data() : llama_get_logits_ith(ctx, seq*n_ctx + first);
+            if (!all_logits) {
+                return 1;
```
Comment (compilade) [inline examples/perplexity/perplexity.cpp:642]: This doesn't match with the return type.

Maybe something like this would be more appropriate?

```suggestion
                return {std::move(tokens), -1, {}, {}};
```

File: common/sampling.cpp [hunk 2]
```diff
@@ -195,6 +195,9 @@ static llama_token llama_sampling_sample_impl(
     llama_token id = 0;
     // Get a pointer to the logits
     float * logits = llama_get_logits_ith(ctx_main, idx);
+    if (!logits) {
+        throw std::runtime_error("llama_get_logits_ith failed");
+    }
```
Comment (compilade) [inline common/sampling.cpp:200]: This could be a `GGML_ASSERT` instead, since `llama_sampling_prepare` was called first and already does this check. Or maybe it could be omitted.

```suggestion
    GGML_ASSERT(logits); // already checked in llama_sampling_prepare 
```

File: common/sampling.cpp [hunk 3]
```diff
@@ -284,6 +287,9 @@ static llama_token_data_array llama_sampling_prepare_impl(
 
     // Get a pointer to the logits
     float * logits = llama_get_logits_ith(ctx_main, idx);
+    if (!logits) {
+        throw std::runtime_error("llama_get_logits_ith failed");
```
Comment (compilade) [inline common/sampling.cpp:291]: >  This change potentially helps prevent the server from crashing.

This exception (like the others added) is never caught, so it will still make the server crash. With `bge-small` and trying to make a chat completion:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  llama_get_logits_ith failed
Aborted (core dumped)
```

Maybe an error code could be returned (using negative numbers, assuming no `llama_token` is negative...), so that the error can then be handled at each call site?

And if it's not handled, maybe make `llama_sampling_accept` crash when it gets a negative `llama_token`?

Or alternatively, slot processing in the server could be wrapped in a `try` block?

I'm not sure what's the cleanest solution here.

Comment (jart) [reply common/sampling.cpp:289]: Doesn't the server have a per request exception handler? Worst case is the client request crashes and 500's, no?

**Comment by jart** [comment]
Thanks for the detailed review!

File: common/sampling.cpp (hunk 3 above)
Comment (compilade) [reply common/sampling.cpp:289]: I tested this before writing, and the server crashes with the error I quoted. On the client side, this is what happens:
```
curl: (18) transfer closed with outstanding read data remaining
```
> Doesn't the server have a per request exception handler?

I would have thought so too, but it might only handle the exceptions of the http library, I'm not sure.

https://github.com/ggerganov/llama.cpp/blob/201cc11afa0a1950e1f632390b2ac6c937a0d8f0/examples/server/server.cpp#L2988-L3001

Comment (jart) [reply common/sampling.cpp]: OK I've pushed some stuff I think should help resolve this.

File: examples/server/server.cpp [hunk 4]
```diff
@@ -2257,6 +2257,9 @@ struct server_context {
 
                 completion_token_output result;
                 const llama_token id = llama_sampling_sample(slot.ctx_sampling, ctx, NULL, slot.i_batch - i);
+                if (id == -1) {
+                    continue; // keep going, don't crash, already logged
```
Comment (compilade) [inline examples/server/server.cpp:2261]: The `server` is still crashing when answering on the `/completion` endpoint with `bge-small`, but now with a different error.
```
GGML_ASSERT: llama.cpp:11006: seq_id < n_tokens && "seq_id cannot be larger than n_tokens with pooling_type == CLS"
```
At least it's no longer due to the logits. Or is it?

Some `fprintf` debugging right before the assertion:

```
llama_set_inputs: seq_id: 1, n_tokens: 5
llama_set_inputs: seq_id: 1, n_tokens: 5
llama_set_inputs: seq_id: 1, n_tokens: 5
llama_set_inputs: seq_id: 1, n_tokens: 5
llama_set_inputs: seq_id: 1, n_tokens: 5
llama_get_logits_ith_fail: invalid logits id 4, reason: no logits
llama_set_inputs: seq_id: 1, n_tokens: 1
GGML_ASSERT: llama.cpp:11007: seq_id < n_tokens && "seq_id cannot be larger than n_tokens with pooling_type == CLS"
```
Something is wrong in the iteration right after `llama_get_logits_ith` fails.

This seems to fix the problem:

```suggestion
                    send_error(slot, "can't get completions out of an embeddings model");
                    slot.cache_tokens.clear();
                    slot.reset();
                    slot.release();
                    slot.i_batch = -1;
                    continue; // continue loop of slots
```

File: examples/speculative/speculative.cpp [hunk 5]
```diff
@@ -457,7 +463,9 @@ int main(int argc, char ** argv) {
                     continue;
                 }
 
-                llama_sampling_sample(drafts[s].ctx_sampling, ctx_dft, NULL, drafts[s].i_batch_dft);
+                if (llama_sampling_sample(drafts[s].ctx_sampling, ctx_dft, NULL, drafts[s].i_batch_dft) == -1) {
+                    return -1;
```
[the remark this answers is not in the document]
Comment (jart) [reply examples/speculative/speculative.cpp:467]: It returns `1` from `main()` functions to `exit(1)`.

[the remark this answers is not in the document]
Comment (compilade) [reply examples/speculative/speculative.cpp:467]: Right. But it's `-1` (negative one) here, so the exit code will be `255`.
